vclib/src/jvmMain/kotlin/at/asitplus/wallet/lib/agent/DefaultCryptoService.kt

```diff
@@ -49,83 +37,89 @@ import kotlin.time.Duration.Companion.days
-        val notBeforeDate = Date.from(Instant.now())
-        val notAfterDate = Date.from(Instant.now().plusSeconds(30.days.inWholeSeconds))
+    private fun generateSelfSignedCertificate(): X509Certificate {
         val serialNumber: BigInteger = BigInteger.valueOf(Random.nextLong().absoluteValue)
-        val issuer = X500Name("CN=DefaultCryptoService")
-        val builder = X509v3CertificateBuilder(
-            /* issuer = */ issuer,
-            /* serial = */ serialNumber,
-            /* notBefore = */ notBeforeDate,
-            /* notAfter = */ notAfterDate,
-            /* subject = */ issuer,
-            /* publicKeyInfo = */ SubjectPublicKeyInfo.getInstance(keyPair.public.encoded)
+        val commonName = "DefaultCryptoService"
+        val notBeforeDate = Clock.System.now()
+        val notAfterDate = notBeforeDate.plus(30, DateTimeUnit.SECOND)
+        val tbsCertificate = TbsCertificate(
+            version = 2,
+            serialNumber = serialNumber.toByteArray(),
+            issuerName = listOf(DistinguishedName.CommonName(Asn1String.UTF8(commonName))),
+            validFrom = Asn1Time(notBeforeDate),
+            validUntil = Asn1Time(notAfterDate),
+            signatureAlgorithm = algorithm,
+            subjectName = listOf(DistinguishedName.CommonName(Asn1String.UTF8(commonName))),
+            publicKey = publicKey
         )
-        val contentSigner: ContentSigner = JcaContentSignerBuilder(JwsAlgorithm.ES256.jcaName).build(keyPair.private)
-        val certificateHolder = builder.build(contentSigner)
-        return certificateHolder.encoded
-    }
-
-    override val jwsAlgorithm = JwsAlgorithm.ES256
-
-    override val coseAlgorithm = CoseAlgorithm.ES256
-
-    override fun toPublicKey() = cryptoPublicKey
-
-    override suspend fun sign(input: ByteArray): KmmResult<ByteArray> =
-        try {
-            val signed = Signature.getInstance(jwsAlgorithm.jcaName).apply {
-                initSign(keyPair.private)
-                update(input)
-            }.sign()
-            KmmResult.success(signed)
-        } catch (e: Throwable) {
-            KmmResult.failure(e)
-        }
+        val signature =
+            runBlocking {
+                runCatching { tbsCertificate.encodeToDer() }
+                    .wrap()
+                    .transform { sign(it) }
+                    .getOrThrow()
+            }
+        return X509Certificate(tbsCertificate, algorithm, signature)
+    }
+
+    override suspend fun sign(input: ByteArray): KmmResult<CryptoSignature> = runCatching {
+        val sig = Signature.getInstance(algorithm.jcaName).apply {
+            this@DefaultCryptoService.algorithm.jcaParams?.let { setParameter(it) }
+            initSign(privateKey)
+            update(input)
+        }.sign()
+        //In Java EC signatures are returned as DER-encoded, RSA signatures however are raw bytearrays
+        if (algorithm.isEc)
+            CryptoSignature.decodeFromDer(sig)
```

Comment:
Would it be a good idea to define a function like `CryptoSignature.parseFromJca(bytes: ByteArray)` to handle that case?

Reply:
probably, but this should then go into a feature request for KMP-Crypto

Reply:
can now be found at `feature/parseJcaSignature` in KMP-Crypto